orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java

```diff
@@ -196,33 +187,55 @@ public void resolveArtifacts(@Nonnull Map pipeline) {
 
   public Artifact resolveSingleArtifact(ExpectedArtifact expectedArtifact, List<Artifact> possibleMatches) {
     List<Artifact> matches = possibleMatches
-      .stream()
-      .filter(expectedArtifact::matches)
-      .collect(toList());
+        .stream()
+        .filter(expectedArtifact::matches)
+        .collect(toList());
     switch (matches.size()) {
       case 0:
         return null;
-      case 1:
-        return matches.get(0);
       default:
-        throw new IllegalStateException(format("Expected artifact %s matches multiple artifacts %s", expectedArtifact, matches));
+        Artifact result = matches.get(0);
+        expectedArtifact.setBoundArtifact(result);
+        return result;
     }
   }
 
-  private ResolveResult resolveExpectedArtifacts(List<ExpectedArtifact> expectedArtifacts, List<Artifact> receivedArtifacts) {
-    ResolveResult result = new ResolveResult();
+  public Set<Artifact> resolveExpectedArtifacts(List<ExpectedArtifact> expectedArtifacts, List<Artifact> receivedArtifacts) {
+    return resolveExpectedArtifacts(expectedArtifacts, receivedArtifacts, null);
+  }
+
+  public Set<Artifact> resolveExpectedArtifacts(List<ExpectedArtifact> expectedArtifacts, List<Artifact> receivedArtifacts, List<Artifact> priorArtifacts) {
+    Set<Artifact> resolvedArtifacts = new HashSet<>();
+    Set<ExpectedArtifact> unresolvedExpectedArtifacts = new HashSet<>();
 
     for (ExpectedArtifact expectedArtifact : expectedArtifacts) {
       Artifact resolved = resolveSingleArtifact(expectedArtifact, receivedArtifacts);
       if (resolved != null) {
-        expectedArtifact.setBoundArtifact(resolved);
-        result.resolvedArtifacts.add(resolved);
+        resolvedArtifacts.add(resolved);
       } else {
-        result.unresolvedExpectedArtifacts.add(expectedArtifact);
+        unresolvedExpectedArtifacts.add(expectedArtifact);
+      }
+    }
+
+    for (ExpectedArtifact expectedArtifact : unresolvedExpectedArtifacts) {
+      Artifact resolved = null;
+      if (expectedArtifact.isUsePriorArtifact() && priorArtifacts != null) {
+        resolved = resolveSingleArtifact(expectedArtifact, priorArtifacts);
+      }
+
+      if (resolved == null && expectedArtifact.isUseDefaultArtifact() && expectedArtifact.getDefaultArtifact() != null) {
+        resolved = expectedArtifact.getDefaultArtifact();
+      }
+
+      if (resolved == null) {
+        throw new IllegalStateException(format("Unmatched expected artifact %s could not be resolved.", expectedArtifact));
+      } else {
+        expectedArtifact.setBoundArtifact(resolved);
```

Comment:
This is really minor, bit it feels to me that `expectedArtifact.setBoundArtifact` and `resolvedArtifacts.add(resolved)` would more logically live a few lines above after `resolved = expectedArtifact.getDefaultArtifact();`.  It also makes it more clear that we are explicitly calling `setBoundArtifact` iff we don't call `resolveSingleArtifact`.

Reply:
Agreed